### PR TITLE
[Kimi-K3] Add multimodal support

### PR DESCRIPTION
**Pull request by Xing-lil**
#### Before submitting

- [ ] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [ ] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

### PR types

Others
### PR changes

Others
### Description

支持 Kimi-K3 的 vision model、数据流预处理、K3 特殊的 XTML 模板

模型层：
- modeling.py：把 vision model 和 language model 拼接到一起，使用 PaddleFleet 中的 k3 multimodel

数据流：
- media_utils.py：底层图像工具，NaViT 风格 resize、padding、patchify、归一化
- vision_processor.py：串起上述工具，输出 pixel_values + grid_thws 给到模型
- processor.py：KimiK3Processor 把图文合到一起
- tokenizer.py：骨架参考 kimi_k25，用 XTML 格式

与其他模型的差异：
- 展开时机：Qwen-VL / GLM-OCR 在 processor 里预展开占位符，K3 延迟到模型内展开，模板需写 expand_mm_tokens=False
- 位置编码：Qwen-VL 用三轴 MRoPE，K3 视觉 token 用普通一维 position id，与文本连续
- 提示词格式：XTML 结构标记

TODO：
- 补齐 multimodal 的 AOA 配置

### CODE REVIEW

File: paddleformers/transformers/kimi_k3/modeling.py [hunk 1]
```diff
@@ -14,6 +14,20 @@
 
 from dataclasses import dataclass
 
+import paddle
+from paddle.distributed import fleet
+from paddle.distributed.fleet.meta_parallel import NoPipelineParallel
+from paddlefleet.models.gpt.gpt_embedding import GPTEmbedding
+from paddlefleet.models.gpt.lm_head import GPTLMHead
+from paddlefleet.models.kimi_k3 import (
```
Comment (risemeup1111) [inline paddleformers/transformers/kimi_k3/modeling.py:22]: P1 当前 `setup.py` 固定的 PaddleFleet 版本中不存在 `paddlefleet.models.kimi_k3`，这个新增顶层导入会让按项目声明安装依赖的 Kimi-K3 模块在导入阶段直接报 `ModuleNotFoundError`。请将依赖提升到实际包含 Kimi-K3 vision 模块的版本，并补充按声明依赖导入该模型的覆盖。

File: paddleformers/transformers/kimi_k3/modeling.py [hunk 2]
```diff
@@ -163,9 +177,246 @@ def __new__(cls, config):
...
+        language_model=NoPipelineParallel(language_model, strategy),
+        media_placeholder_token_id=config.media_placeholder_token_id,
+        pad_token_id=getattr(config, "pad_token_id", None),
+        ignore_index=getattr(config, "ignore_index", -100),
+    )
+    model.config_to_save = config
+    return FleetKimiK3ForConditionalGeneration(config, model, criterion)
+
+
+class KimiK3ForConditionalGeneration(KimiK3PretrainedModel):
```
Comment (risemeup1111) [inline paddleformers/transformers/kimi_k3/modeling.py:813]: P1 新增的 `KimiK3ForConditionalGeneration` 没有 `_gen_aoa_config`，而 `PretrainedModel.from_pretrained` 默认使用 `flex_checkpoint`，会在现有检查处直接抛出 `RuntimeError`，因此官方 HF 多模态 checkpoint 无法通过默认入口加载。请补齐覆盖 text、vision、projector 及 wrapper state-dict key 的 AoA 映射，并增加实际 load/save 回归。

File: paddleformers/transformers/kimi_k3/modeling.py [hunk 3]
```diff
@@ -163,9 +177,246 @@ def __new__(cls, config):
...
+    def forward(self, dict_args=None, **kwargs):
+        """Run the multimodal model and return the scalar training loss.
+
+        Accepts either a single ``dict_args`` mapping or plain keyword arguments,
+        because ``Trainer.compute_loss`` calls ``model(**inputs)`` for models it
+        does not recognise as a Fleet ``GPTModel``.
+        """
+        if dict_args is None:
+            dict_args = kwargs
+        logits = self.model(dict_args)
```
Comment (risemeup1111) [inline paddleformers/transformers/kimi_k3/modeling.py:762]: P1 这个 `ForConditionalGeneration` forward 无条件进入 criterion。标准 `generate()` 不会传 `labels`，首次 forward 就会在 `sft_calculate_loss` 访问 `labels.shape` 失败，而且这里返回的是 loss 而不是 generation loop 需要的 logits；当前也没有为后续 decode step 清除 `pixel_values/grid_thws`，即使加上 labels 判断，通用 generation 仍会对不含 media placeholder 的单 token 再次执行图像 merge。请补齐 prefill/decode 的 logits 与缓存输入契约，并加入含图 `generate()` 回归。

File: paddleformers/transformers/kimi_k3/vision_processor.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+            data = {
+                "pixel_values": pixel_values,
+                "grid_thws": grid_thws,
+            }
+
+        else:
+            data = {}
+
+        return BatchFeature(data=data, tensor_type="pd")
```
Comment (risemeup1111) [inline paddleformers/transformers/kimi_k3/vision_processor.py]: P2 `return_tensors` 在这里被忽略：前面已经无条件把数组转成 Paddle Tensor，最终又固定为 `tensor_type="pd"`，所以 `preprocess(..., return_tensors="np")` 或 `None` 仍返回 Paddle Tensor，与方法签名和文档不符。请按请求类型保留或转换数据，或明确校验只支持 `pd`，并补充返回类型测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [modeling.py:22](https://github.com/PaddlePaddle/PaddleFormers/pull/4834#discussion_r3718017344) | P1 | 🚧 |
| 2 | [modeling.py:401](https://github.com/PaddlePaddle/PaddleFormers/pull/4834#discussion_r3718019698) | P1 | 🚧 |
| 3 | [modeling.py:350](https://github.com/PaddlePaddle/PaddleFormers/pull/4834#discussion_r3718020853) | P1 | 🚧 |
| 4 | [vision_processor.py:179](https://github.com/PaddlePaddle/PaddleFormers/pull/4834#discussion_r3718021906) | P2 | ✅ |
| 5 | [modeling.py:282](https://github.com/PaddlePaddle/PaddleFormers/pull/4834#discussion_r3718439134) | P1 | ✅ |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/kimi_k3/modeling.py (hunk 3 above)
Comment (risemeup1111) [reply paddleformers/transformers/kimi_k3/modeling.py:762]: 本次提交已修复 `labels=None` 时首次 forward 直接进入 criterion 的问题，并会返回 logits。但 `KimiK3VLModel.forward` 仍只在 `modeling.py:325` 返回 logits tensor，没有接收或返回 `past_key_values`，也没有 `prepare_inputs_for_generation` 或 prefill/decode 分支来移除 `pixel_values/grid_thws`；后续 decode step 仍会重跑完整视觉融合，通用 `generate()` 的缓存输入契约仍未满足，因此该 P1 保持未解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: paddleformers/transformers/kimi_k3/modeling.py (hunk 1 above)
Comment (Xing-lil) [reply paddleformers/transformers/kimi_k3/modeling.py:22]: CI 没问题

File: paddleformers/transformers/kimi_k3/modeling.py (hunk 2 above)
Comment (Xing-lil) [reply paddleformers/transformers/kimi_k3/modeling.py:813]: AOA 后续会完善

File: paddleformers/transformers/kimi_k3/modeling.py [hunk 5]
```diff
@@ -163,9 +177,247 @@ def __new__(cls, config):
...
+    def forward(self, dict_args):
+        """Embed, fuse the visual tokens, then run the text backbone.
+
+        Media inputs make the sequence longer, so ``attention_mask`` / ``labels``
+        / ``position_ids`` are rewritten in ``dict_args`` for the caller to read
+        back after this returns.
+        """
+        input_ids = dict_args["input_ids"]
+        pixel_values = dict_args.get("pixel_values", None)
+        grid_thws = dict_args.get("image_grid_thw", None)
```
Comment (risemeup1111) [inline paddleformers/transformers/kimi_k3/modeling.py:694]: P1 当前模型只从 `dict_args` 读取 `image_grid_thw`，但同一提交的 `KimiK3VisionProcessor.preprocess` 返回 `grid_thws`，`KimiK3Processor.model_input_names` 也声明 `grid_thws`。将处理器输出直接传给该模型会在 `pixel_values != None` 时触发缺少 `image_grid_thw` 的异常，导致标准多模态输入链路不可用。请统一键名（或兼容两个键），并补充 processor→model 回归。

File: paddleformers/transformers/kimi_k3/modeling.py (hunk 3 above)
Comment (risemeup1111) [reply paddleformers/transformers/kimi_k3/modeling.py:762]: 本次提交改为明确拒绝 `labels=None`，并注明 generation 尚未实现。这避免了原先首次 forward 的隐式 `labels.shape` 错误，但含图 `generate()` 仍不可用，且 wrapper 仍无 KV-cache/prefill-decode 契约；该 P1 保持未解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: paddleformers/transformers/kimi_k3/modeling.py (hunk 1 above)
Comment (risemeup1111) [reply paddleformers/transformers/kimi_k3/modeling.py:22]: CI 状态不改变该兼容性结论：当前 `setup.py` 仍固定同一 PaddleFleet 版本，提交中也没有证明该版本提供 `paddlefleet.models.kimi_k3`。请更新声明依赖并补充按声明依赖导入的回归；该 P1 保持未解决。

File: paddleformers/transformers/kimi_k3/modeling.py (hunk 2 above)
Comment (risemeup1111) [reply paddleformers/transformers/kimi_k3/modeling.py:813]: 收到。当前提交仍未提供 `_gen_aoa_config` 或 `_gen_inv_aoa_config`，也没有 load/save 回归验证，因此官方多模态 checkpoint 的默认 FlexCheckpoint 入口仍不可用；该 P1 保持待修复。

File: paddleformers/transformers/kimi_k3/modeling.py (hunk 5 above)
Comment (risemeup1111) [reply paddleformers/transformers/kimi_k3/modeling.py:694]: 本次提交已统一该输入契约：`KimiK3VisionProcessor`、`KimiK3Processor` 和数据插件现在都使用 `image_grid_thw`，与 `KimiK3VLModel.forward` 的读取键一致；该 P1 已解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: paddleformers/transformers/kimi_k3/modeling.py (hunk 2 above)
Comment (risemeup1111) [reply paddleformers/transformers/kimi_k3/modeling.py:813]: 本次合入仅补上了 text backbone 的 AoA：`_gen_aoa_config` / `_gen_inv_aoa_config` 的语句只覆盖 `language_model.*`，未映射 `model.visual` 下的 vision tower/projector；同时 `_build_vl_model` 返回的 `FleetKimiK3ForConditionalGeneration` 未像 `_build_text_model` 一样绑定这两个方法，因此 `save_pretrained(..., save_checkpoint_format="flex_checkpoint")` 仍会在实例检查处失败。请为多模态 wrapper 补齐双向映射并增加实际 load/save 回归；该 P1 保持未解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by From00 — APPROVED** [review]
LGTM
